game/scripts/vscripts/units/shielder.lua

```diff
@@ -0,0 +1,212 @@
+    DeathHandler(state, keys)
+  end)
+
+  unit:SetIdleAcquire(false)
+  unit:SetAcquisitionRange(0)
+
+  return {
+    onDeath = state.deathEvent.listen
+  }
+end
+
+function HurtHandler (state, keys)
+  if state.state == 1 then
+    DebugPrint('Checking boss agro...')
+    DebugPrintTable(keys)
+
+    state.currentDamage = state.currentDamage + keys.damage
+
+    if state.currentDamage > state.agroDamage then
+      Agro(state, EntIndexToHScript(keys.entindex_attacker))
+      state.currentDamage = 0
+    end
+  elseif state.state == 2 then
+  end
+end
+
+function GiveItemToWholeTeam (item, teamId)
+  for playerId = 0,19 do
+    if PlayerResource:GetTeam(playerId) == teamId and PlayerResource:GetPlayer(playerId) ~= nil then
+      local player = PlayerResource:GetPlayer(playerId)
+      local hero = player:GetAssignedHero()
+
+      hero:AddItemByName(item)
+    end
+  end
+end
+
+function DeathHandler (state, keys)
+  DebugPrint('Handling death of boss ' .. state.tier)
+  state.state = 4
+
+  state.handle = nil
+  local killer = EntIndexToHScript(keys.entindex_attacker)
+  local teamId = killer:GetTeam()
+
+  state.deathEvent.broadcast(keys)
+
+  if teamId == 2 then
+    team = 'good'
+  elseif teamId == 3 then
+    team = 'bad'
+  else
+    return
+  end
+
+  GiveItemToWholeTeam("item_upgrade_core", teamId)
+
+  for playerId = 0,19 do
+    if PlayerResource:GetTeam(playerId) == teamId and PlayerResource:GetPlayer(playerId) ~= nil then
+      local player = PlayerResource:GetPlayer(playerId)
+      local hero = player:GetAssignedHero()
+
+      if hero then
+        if not hero.hasFarmingCore then
+          hero:AddItemByName("item_farming_core")
+          hero.hasFarmingCore = true
+        elseif not hero.hasReflexCore then
+          hero:AddItemByName("item_reflex_core")
+          hero.hasReflexCore = true
+        end
+      end
+    end
+  end
+end
+
+function Agro (state, target)
+  if state.customAgro then
+    DebugPrint('Running custom agro ai')
+    return
+  end
+
+  Timers:CreateTimer(1, function ()
+    if state.state == 4 then
+      return
+    end
+
+    if not Think(state) or state.state == 1 then
+      DebugPrint('Stopping think timer')
+      return
+    end
+    return 1
+  end)
+  state.state = 2
+  state.agroTarget = target
+
+  state.handle:SetIdleAcquire(true)
+  state.handle:SetAcquisitionRange(128)
+
+  ExecuteOrderFromTable({
+    UnitIndex = state.handle:entindex(),
+    OrderType = DOTA_UNIT_ORDER_ATTACK_TARGET,
+    --OrderType = DOTA_UNIT_ORDER_ATTACK_MOVE,
+    Position = target:GetAbsOrigin(),
+    Queue = 0,
+  })
+  ExecuteOrderFromTable({
+    UnitIndex = state.handle:entindex(),
+    OrderType = DOTA_UNIT_ORDER_ATTACK_TARGET,
+    --OrderType = DOTA_UNIT_ORDER_ATTACK_MOVE,
+    Position = state.origin,
+    Queue = 1,
+  })
+end
+
+function ShielderThink (state, target)
+  Think(state)
+end
+
+function Think (state)
+  local distance = (state.handle:GetAbsOrigin() - state.origin):Length()
```

Comment:
You nil out handle in your OnDeath function, thinkers still run after death (for cleanup or post-death animations).... You should
```lua
if state.handle == nil then
  return 0
end
```